### DISCUSSION TITLE
Toggles states aren't saved

### DISCUSSION THREAD

**Discussion by hadiidbouk** [pointfreeco/swift-composable-architecture#2940] [Q&A]
I was able to reproduce an issue we had in one of our projects where we have toggles in a `SettingsView` and if we change the value of one of these toggles they won't be saved as a state in memory. The state of Settings keeps resetting even when I am trying to use the same state over and over again.

Here is a video that shows the issue:
https://github.com/pointfreeco/swift-composable-architecture/assets/19314956/524196e5-0794-4206-bf43-95b7833d9c9f

The code:

```swift
import ComposableArchitecture
import SwiftUI

@Reducer
struct ContentFeature {
	@Reducer
	struct Path {
		@ObservableState
		enum State {
			case settings(SettingsFeature.State)
		}
		enum Action {
			case settings(SettingsFeature.Action)
		}
		var body: some ReducerOf<Self> {
			Scope(state: \.settings, action: \.settings) {
				SettingsFeature()
			}
		}
	}

	@ObservableState
	struct State {
		var path = StackState<Path.State>()
		let settings = SettingsFeature.State()
	}

	enum Action {
		case path(StackAction<Path.State, Path.Action>)
	}

	var body: some ReducerOf<Self> {
		Reduce { state, action in
			switch action {
			default:
				return .none
			}
		}
		.forEach(\.path, action: \.path) {
			Path()
		}
	}
}

struct ContentView: View {
	@Bindable var store: StoreOf<ContentFeature>

    var body: some View {
		NavigationStack(path: $store.scope(state: \.path, action: \.path)) {
			NavigationLink(
				"Go to settings",
				state: ContentFeature.Path.State.settings(store.settings)
			)
		} destination: { store in
			switch store.state {
			case .settings:
				if let store = store.scope(state: \.settings, action: \.settings) {
					SettingsView(store: store)
				}
			}
		}
    }
}

//MARK: Settings
@Reducer
struct SettingsFeature {
	@ObservableState
	struct State {
		var isDebugModeEnabled = false
		var isAppReadyForProd = false
	}
	enum Action {
		case debugModeToggleChanged(Bool)
		case appReadyToggleChanged(Bool)
		case closeButtonTapped
	}

	@Dependency(\.dismiss) private var dismiss

	var body: some ReducerOf<Self> {
		Reduce { state, action in
			switch action {
			case .debugModeToggleChanged(let isEnabled):
				state.isDebugModeEnabled = isEnabled
				return .none
			case .appReadyToggleChanged(let isEnabled):
				state.isAppReadyForProd = isEnabled
				return .none
			case .closeButtonTapped:
				return .run { _ in
					await self.dismiss()
				}
			}
		}
	}
}

struct SettingsView: View {
	@Bindable var store: StoreOf<SettingsFeature>

	var body: some View {
		List {
			Toggle(isOn: $store.isDebugModeEnabled.sending(\.debugModeToggleChanged)) {
				Text("Enable Debug Mode")
			}
			.toggleStyle(.switch)

			Toggle(isOn: $store.isAppReadyForProd.sending(\.appReadyToggleChanged)) {
				Text("Is App Ready For Production?")
			}
			.toggleStyle(.switch)

			Button("Close") {
				store.send(.closeButtonTapped)
			}
		}
	}
}

#Preview {
	ContentView(
		store: Store(initialState: ContentFeature.State()) {
			ContentFeature()
		}
	)
}
```

I want to preserve the settings state after the second navigation, how can I do that?

**Comment by kamcma** [comment]
This is entirely expected. When you push `SettingsFeature.State` onto the navigation stack here:
```swift
NavigationLink(
  "Go to settings",
  state: ContentFeature.Path.State.settings(store.settings)
)
```
that value, `store.settings`, is copied _by value_. It immediately become disconnected from the original copy of that value held as an instance by `ContentFeature.State`, and it will not mutate that original copy. Then when that state is popped off the stack, it is lost unless you do something to capture it and hold it elsewhere.

Now, this is a _very_ timely issue for you to have. There is currently a TCA beta for an upcoming feature called shared state (#2857) that is a very ergonomic way to pass state around with reference-y semantics like you want to. I would highly recommend you watch [the (still-in-progress) video series](https://www.pointfree.co/collections/composable-architecture/sharing-and-persisting-state) about this feature, and in particular, in [the solution-part-2 episode](https://www.pointfree.co/collections/composable-architecture/sharing-and-persisting-state/ep270-shared-state-the-solution-part-2), they demonstrate passing shared state around different levels of a navigation stack, exactly what it looks like you are trying to accomplish.

**Reply by kamcma** [reply]
Your `SettingsFeature` could also simplified/shortened by leveraging [the binding tools TCA provides](https://pointfreeco.github.io/swift-composable-architecture/1.9.0/documentation/composablearchitecture/bindings).
